hydra/core/callbacks.py

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+import warnings
+from typing import Any, List
+
+from omegaconf import DictConfig
+
+from hydra._internal.callback import Callback
+from hydra.core.singleton import Singleton
+from hydra.core.utils import JobReturn
+from hydra.utils import instantiate
+
+
+class Callbacks(metaclass=Singleton):
```

Comment:
Why should this be a singleton?
Looking at how you are using it, I suspect there isn't really a reason for this. making it a normal member of the Hydra object should work just fine.

(You will need to pass it explicitly into run_job()).

Reply:
I've considered passing it explicitly: If we pass in Callbacks to every run_job that means we will have pass it from sweepers then on to all the launchers. 
Meaning we will probably have to change all the sweeper and launcher code. 
for example, we'd have to change [here](https://github.com/facebookresearch/hydra/blob/37cdb2daf75c9a0af37952122405af8f83449772/plugins/hydra_joblib_launcher/hydra_plugins/hydra_joblib_launcher/_core.py#L43-L48) for the joblib launcher

Another alternative is to init callbacks every time we need then, but that doesn't seem as clean (we will have to init callbacks twice 
during the life of a Hydra application)

After trying out a few alternative ways, I think having this as a Singleton is the cleanest way to get around the above. 

What do you think?

Reply:
I think we should it init it once, and pass it down to plugins somehow (e.g. Sweeper -> Launcher -> run_job). Using a singleton here is just accruing technical debt.

What do you think about the following:

Introduce a HydraContext object, it can contain ConfigLoader and the callbacks for now.
We now need to make sure HydraContext is propagated in.

This can either be a breaking change in plugins:
e.g. a change to the signature of setup(), in this case, we can release dev release of existing plugins for Hydra 1.1 users) (and also update the dependencies of Hydra 1.0 plugins to be sure they are marked as Hydra 1.0 only).

Alternatively we can try to make it in steps while maintaining backward compatibility.
In this case I would like to see a plan of how we do it (what happens at each Hydra version).

I am leaning toward just ripping the band-aid and go with option 1, but open to option 2 as well.